### PR TITLE
fix: Eliminate race conditions in reconnect feature.

### PR DESCRIPTION
**Pull request by mturoci**
Reconnect feature stabilization.

This PR adds R/W synchronization to client state shared across multiple goroutines, preventing race conditions. A simple state machine is used for the synchronization.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -90,29 +98,44 @@ func (c *Client) refreshToken() error {
...
-				}
+		c.lock.Lock()
+		defer c.lock.Unlock()
+		if c.state != STATE_DISCONNECT {
+			return
+		}
+		// This defer runs to completion. If the client drops, reconnects and drops out again, ignore first drop timeout.
+		timeoutID := STATE_TIMEOUT + c.addr
+		c.state = timeoutID
+		c.lock.Unlock()
```
Comment (geomodular) [inline client.go:117]: This line will make `c.lock` to unlock twice (due to `defer`).

File: client.go [hunk 2]
```diff
@@ -90,29 +98,44 @@ func (c *Client) refreshToken() error {
 	return nil
 }
 
+func (c *Client) setState(newState string) {
+	c.lock.Lock()
+	c.state = newState
```
Comment (geomodular) [inline client.go:103]: Is it necessary to set a variable in a locked environment?

File: client.go [hunk 3]
```diff
@@ -33,10 +34,16 @@ const (
...
 var (
-	newline       = []byte{'\n'}
-	notFoundMsg   = []byte(`{"e":"not_found"}`)
-	disconnectMsg = []byte(`{"data": {"":{"@system":{"client_disconnect":true}}}}`)
-	clearStateMsg = []byte(`{"c":1}`)
+	newline          = []byte{'\n'}
+	notFoundMsg      = []byte(`{"e":"not_found"}`)
+	disconnectMsg    = []byte(`{"data": {"":{"@system":{"client_disconnect":true}}}}`)
+	clearStateMsg    = []byte(`{"c":1}`)
+	STATE_CREATED    = "CREATED"
```
Comment (geomodular) [inline client.go:41]: The idiomatic way to represent a state (enumeration) would be a `const` value (not `var`) generated by `iota`(are those values in variables needed?).

File: client.go (hunk 1 above)
Comment (mturoci) [reply client.go:117]: Wouldn't the deferred unlock run only after `select` statement is finished?

File: client.go (hunk 2 above)
Comment (mturoci) [reply client.go:103]: Yes, because the variable is accessed across different goroutines. Better ideas welcome ofc.

File: client.go (hunk 3 above)
Comment (mturoci) [reply client.go:41]: This was easier to debug, but since it works, can be refactored into `iota`.

File: client.go (hunk 1 above)
Comment (geomodular) [reply client.go:117]: Ah, sorry, now I see what you meant!

**Review by geomodular — APPROVED** [review]
LGTM, thanks Martin!

**Pull request merged**